website/blog/2025/05-05-arbiter-type-class/index.md

```diff
@@ -0,0 +1,280 @@
+
+```scala
+class GenericPriorityArbiter[A <: Data](
+    nClients: Int,
+    clientType: A
+  )(
+    requestFn: A => Bool,
+    grantFn: (A, Bool) => Unit) extends Module {
+  val clients = IO(Vec(nClients, Flipped(clientType)))
+
+  val requests = clients.map(requestFn(_))
+  val granted = PriorityEncoderOH(requests)
+
+  for ((client, grant) <- clients.zip(granted)) {
+    grantFn(client, grant)
+  }
+}
+```
+
+You may notice this looks quite similar to the original `PriorityArbiter` in its implementation.
+It uses two parameter lists in order to help the Scala type inferencer derive the types of the functions of the type of the client--we could do this with one parameter list but then it would require explicitly passing the type of the client.
+
+Now we can use it for both our `ArbiterClient` and `Decoupled` interfaces.
+
+```scala
+val clients1 = IO(Vec(4, Flipped(new ArbiterClient)))
+val arbiter1 = Module(
+  new GenericPriorityArbiter(4, new ArbiterClient)(_.request, (c, g) => c.grant := g)
+)
+arbiter1.clients :<>= clients1
+
+val clients2 = IO(Vec(4, Flipped(Decoupled(UInt(8.W)))))
+val arbiter2 = Module(
+  new GenericPriorityArbiter(4, Decoupled(UInt(8.W)))(_.valid, (d, g) => d.ready := g)
+)
+arbiter2.clients :<>= clients2
+```
+
+This is still a bit clunky--we have to pass two additional arguments to the arbiter.
+
+Another thing to consider is that these additional arguments are the same for all instances of a given type.
+For example, if we were to arbitrate another set of `Decoupled` interfaces, we would have to pass the same functions again:
+
+```scala
+val clients3 = IO(Vec(4, Flipped(Decoupled(UInt(8.W)))))
+val arbiter3 = Module(
+  new GenericPriorityArbiter(4, Flipped(Decoupled(UInt(8.W))))(_.valid, (d, g) => d.ready := g)
+) // the two function arguments are the same as above.
+arbiter2.clients :<>= clients3
+```
+
+## Introducing a Type Class
+
+To clean this up even more, we can introduce a type class that captures the "arbitrable" pattern:
+
+```scala
+trait Arbitrable[A] {
+  def request(a: A): Bool
+  def grant(a: A, value: Bool): Unit
+}
+```
+
+Effectively, we have taken the two arguments to the arbiter and turned them into methods on the type class.
+This looks similar to the proposed object-oriented version of `Arbitrable` above, but note how it is parameterized by the type of the client and accepts the client as an argument.
+
+We can then provide instances of this type class for specific types. For example, for `ArbiterClient` and `Decoupled`:
+
+```scala
+class ArbiterClientArbitrable extends Arbitrable[ArbiterClient] {
+  def request(a: ArbiterClient) = a.request
+  def grant(a: ArbiterClient, value: Bool) = a.grant := value
+}
+
+class DecoupledArbitrable[T <: Data] extends Arbitrable[DecoupledIO[T]] {
+  def request(a: DecoupledIO[T]) = a.valid
+  def grant(a: DecoupledIO[T], value: Bool) = a.ready := value
+}
+```
+
+Then, we can refactor the arbiter to use the type class:
```

Comment:
```suggestion
Then, we can refactor the arbiter to use the type class so that we're getting one 'package' of functions for type A, not having to pass them individually:
```